### ISSUE TITLE
Validating Requests based on models

### ISSUE THREAD

**Issue by zanona** [awslabs/aws-apigateway-swagger-importer#175]
Hi guys,
Sorry if this is out of place for this repository.
But I have noticed that by setting Model definitions in my swagger file and then making use of features as `required`, `minimum`, `maximum`, `enum`, etc.

When importing to API Gateway, those are created as a model, however the requests are not validated at all and the user can simply post anything to a request and that will still go through?

I am not sure If I am getting this right, but without having the validation for models, etc, following swagger specification doesn't seem very helpful?

Since I am very new to this, it might happen that I am doing something completely wrong?

Thank you

**Comment by kahouieong** [comment]
> When importing to API Gateway, those are created as a model, however the requests are not validated at all and the user can simply post anything to a request and that will still go through?

Currently, API Gateway doesn't do any validation even it is defined in the model. The model is used by generating a SDK for your API. Without using a SDK, the user can pass anything to a request and it will go through.

> I am not sure If I am getting this right, but without having the validation for models, etc, following swagger specification doesn't seem very helpful?

You still can define your API structure, your integration endpoint and parameters mapping. It is also easier to manage in source control system.

> Since I am very new to this, it might happen that I am doing something completely wrong?

API Gateway provides a test invoke feature. It allows you to test your API setting without doing an actual API deployment.

**Comment by zanona** [comment]
thanks for the explanation @kahouieong that was helpful?
Are there any plans for adding the data validation layer directly on API Gateway so it's not needed to reach a lambda function until all the data is correctly sent? At the moment it feels a bit cumbersome having to duplicate all the information present on my Swagger models to the lambda itself in regards data validation, after all Swagger format provides us with many features coming from JSON Schema that would allow data validation done directly in API Gateway such `enum`, `minimum`, `maximumLength`, etc. So it would be great to know if you guys are thinking about this, taking out the "plumbing" part away from the development process which would be really helpful.

Another question related to that is, even after generating the SDKs two different things happened taking this model example:

``` yaml
definitions:
  User:
    type: object
    properties:
      name:
        type: string
      email:
        type: string
        format: email
      birthday:
        type: string
        format: date'
      friends_count:
        type: integer
        maximum: 6
        minimum: 1
        default: 2
      color:
        type: string
        enum:
          - blue
          - red
          - green
    required:
      - name
      - email
      - color
```

### Javascript SDK

It doesn't generate any models whatsoever, even though both request and responses models have the `User` model linked to it. So in this case, Model has no relevance to the whole system at all

### iOS SDK

The `User` class only declare properties for the definition parameters such as (name, email, birthday, etc) with the correct types, however, none of the validation is put in place even on the SDK, ignoring required fields, default values, max/min and enum values.

So after spending a considerable amount of time trying to get it right, I was quite surprised to see that even the SDKs which were responsible for handling the models are not doing it reliably?

In the end I continue to wonder if Models have any kind of relevance, at least on this moment, when using API Gateway since most part of the data is mostly skipped?

Hopefully this will change in the future?

**Comment by kahouieong** [comment]
We appreciate your feedbacks. I have forwarded this to API Gateway team.

**Comment by zanona** [comment]
thanks @kahouieong, you are appreciated! I see a lot of potential on this.

**Comment by rpgreen** [comment]
@zanona this is definitely on our roadmap - thanks for the feedback!

**Comment by romain-nio** [comment]
it sounds to be a great feature on my side too !

**Comment by romain-nio** [comment]
Is there a lib (in node.js for example) that can check a JSON payload is valid, according to a Swagger objects definition ? (the swagger API definition can be hosted on S3 for example).

If the answer is yes, that can be a good way to automatically check that a request is valid, without editing the lambda

**Comment by zanona** [comment]
@romain-nio I have built something based on [sway](https://github.com/apigee-127/sway) already, but that's really early stage and it works for my purposes at the moment, however also check a look on our latest discussion here: https://github.com/apigee-127/sway/issues/68 which might be informative. at the moment requests can get validated by simply using sway api, although the library is a bit heavy at the moment. but that can be but put as a middleware in between lambda and your app.

Although, back to the original issue here, there are a serious need for this to be implemented directly at API Gateway so all the plumbing work wouldn't be necessary at our end.

**Comment by rpgreen** [comment]
Swagger/OpenAPI import is now generally available in the API Gateway [REST API](http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-import-api.html), the AWS [CLI](http://docs.aws.amazon.com/cli/latest/reference/apigateway/put-rest-api.html) and all AWS [SDKs](https://aws.amazon.com/tools). You can also import and export Swagger definitions using the API Gateway [console](https://console.aws.amazon.com/apigateway/home). This release addresses many of the open issues and feedback in this repository.

I would encourage you to migrate your workflow to the standard AWS tools. aws-apigateway-importer will receive minimal support from the API Gateway team going forward.

Any feedback or issues with the new importer should be directed to the official API Gateway [forums](https://forums.aws.amazon.com/forum.jspa?forumID=199).

Thanks,
Ryan

**Comment by binoculars** [comment]
I was also very surprised that API Gateway doesn't validate the input--I could've sworn this was advertised somewhere before. It's easy enough to do it in Lambda, but I would love to see it implemented in API Gateway as an option. At least when using the service proxy integration request with DynamoDB, I can use a ConditionExpression for now. Thanks and keep up the great work!

**Comment by iWebi** [comment]
Is there any date in plan for request payload validation for API Gateway?

**Comment by amasare** [comment]
Any dates yet?

**Comment by rajeget** [comment]
Again Any Dates AWS ?

**Comment by binoculars** [comment]
AWS usually doesn't provide dates for new features/enhancements.

**Comment by khalidx** [comment]
Has the swagger input request model validation been added to the AWS API gateway? I've searched far and wide and it seems that without additional middleware a swagger import into the gateway doesn't include this feature.

**Comment by aabashkin** [comment]
+1, this would be great

**Comment by cwbeck** [comment]
Whoever designed the API Gateway should be fired. This is complete joke of a platform. You might want to at least change the wording in your docs AWS...

> “API Gateway maps a payload according to a mapping template. A model is useful, but not required, to generate a template. However, models are necessary for generating strongly typed SDK of your API. They can also be useful to validate a payload."

Angry I wasted so much time on this.

**Comment by rpgreen** [comment]
This feature has been released. Thanks for your patience on this. Please post feedback in the official AWS Forums. -Ryan

https://aws.amazon.com/blogs/compute/how-to-remove-boilerplate-validation-logic-in-your-rest-apis-with-amazon-api-gateway-request-validation/

**Comment by Shockolate** [comment]
@rpgreen 
Read your article, very exciting!
However, the example swagger file (JSON) is Invalid swagger.... It's riddled with errors....

Here's a valid version of the swagger file.
```yaml
swagger: '2.0'
info:
  version: 1.0.0
  title: Fixed Swagger Request Validation Demo - rpgreen@amazon.com
schemes:
  - https
produces:
  - application/json
x-amazon-apigateway-request-validators:
  full:
    validateRequestBody: true
    validateRequestParameters: true
  body-only:
    validateRequestBody: true
    validateRequestParameters: false
x-amazon-apigateway-request-validator: full
paths:
  /orders:
    post:
      x-amazon-apigateway-request-validator: body-only
      parameters:
        - in: body
          name: CreateOrders
          required: true
          schema:
            $ref: '#/definitions/CreateOrders'
      responses:
        '200':
          schema:
            $ref: '#/definitions/Message'
          description: 'Valid'
        '400':
          schema:
            $ref: '#/definitions/Message'
          description: 'BadRequest'
      x-amazon-apigateway-integration:
        responses:
          default:
            statusCode: '200'
            responseTemplates:
              application/json: '{"message" : "Orders successfully created"}'
        requestTemplates:
          application/json: '{"statusCode": 200}'
        passthroughBehavior: never
        type: mock
    get:
      parameters:
        - in: header
          name: Account-Id
          required: true
          type: string
        - in: query
          name: type
          required: false
          type: string
      responses:
        '200':
          schema:
            $ref: '#/definitions/Orders'
          description: 'Valid'
        '400':
          schema:
            $ref: '#/definitions/Message'
          description: 'BadRequest'
      x-amazon-apigateway-integration:
        responses:
          default:
            statusCode: '200'
            responseTemplates:
              application/json: |-
                [{"order-id" : "qrx987",
                   "type" : "STOCK",
                   "symbol" : "AMZN",
                   "shares" : 100,
                   "time" : "1488217405",
                   "state" : "COMPLETED"
                },
                {
                   "order-id" : "foo123",
                   "type" : "STOCK",
                   "symbol" : "BA",
                   "shares" : 100,
                   "time" : "1488213043",
                   "state" : "COMPLETED"
                }
                ]
        requestTemplates:
          application/json: '{"statusCode": 200}'
        passthroughBehavior: never
        type: mock
definitions:
  CreateOrders:
    title: Create Orders Schema
    type: array
    minItems: 1
    items:
      $ref: '#/definitions/Order'
  Orders:
    type: array
    title: Get Orders Schema
    items:
      type: object
      properties:
        order_id:
          type: string
        time:
          type: string
        state:
          type: string
          enum:
            - PENDING
            - COMPLETED
        order:
          $ref: '#/definitions/Order'
  Order:
    type: object
    title: Schema for a single Order
    required:
      - account-id
      - type
      - symbol
      - shares
      - details
    properties:
      account-id:
        type: string
        pattern: '[A-Za-z]{6}[0-9]{6}'
      type:
        type: string
        enum:
          - STOCK
          - BOND
          - CASH
      symbol:
        type: string
        minLength: 1
        maxLength: 4
      shares:
        type: number
        minimum: 1
        maximum: 1000
      details:
        type: object
        required:
          - limit
        properties:
          limit:
            type: number
  Message:
    type: object
    properties:
      message:
        type: string
```

**Comment by binoculars** [comment]
Is there a way to do this with CloudFormation (not using Swagger)

**Comment by rpgreen** [comment]
@Shockolate thanks! You are correct in that Swagger editor shows errors. Swagger parser and API Gateway don't care about these however. I've updated this in the repo at https://github.com/rpgreen/apigateway-validation-demo/blob/master/validation-swagger.json

**Comment by rpgreen** [comment]
@binoculars not yet but this should be coming soon. AWS SDK/CLI support should be released this week.

**Comment by Shockolate** [comment]
Is there a way to add a more verbose error message to the response to the client? IMO there is a better integration experience if the client knows _why_ the request is bad.

Will probably cross post this on the AWS Forums or AWS Enterprise Support

**Comment by rpgreen** [comment]
This is true in some cases, but not all API developers want to expose specific error messages. With that said, the team is tracking a separate item to generally support customizable error responses, so stay tuned for that.

**Comment by sidashah** [comment]
I can't get Access-Control-Allow-Origin Header in the response from API Gateway when the response is "Invalid request body". Any workaround to this?

**Comment by lazdmx** [comment]
@rpgreen Hi! Are there any way to modify response body and/or get a reason of error in case of schema validation failure?

**Comment by Shockolate** [comment]
@lazutkin 
They're working on it (read a couple messages up) 
:)

**Comment by sid88in** [comment]
@rpgreen THIS IS GREAT! 👍  I spent whole day figuring out how to do this and stumbled upon your post later in the day. Thanks for putting this together. 

I have a quick question - In one of my test cases, looks like I can post any extra json field that is not present in the swagger model. For instance, in your above example, the following post call works (I have added a field "something" in the POST call. Ideally I should get a 400 {"message": "Invalid request body"} but I get { "message": "Orders successfully created"}

Test case 1:
```
[  
   { 
      "account-id":"abcdef123456",
      "type":"STOCK",
      "symbol":"AMZN",
      "shares":100,
      "something" : "something",
      "details":{  
         "limit":1000
      }
   }
] 
```

Test case 2:
```
[  
   { 
      "account-id":"abcdef123456",
      "type":"STOCK",
      "symbol":"AMZN",
      "shares":100,
      "details":{  
         "limit":1000,
         "something" : "something"
      }
   }
]

```
Is there any way we can stop the POST call from returning 200 OK. Any thoughts?

Another use case, related to the same topic:

I created a date-time field (record_ts) to get the current timestamp of the POST call made by the user (Idea is that user doesn't need to pass the record_ts, it can automatically be generated on fly and included in the json record) The field looks like:
```
properties:
      record_ts:
        type: string
        format: date-time
        default: now()
```

So, in the final POST call the user would not send record_ts but because our model specifies it, record_ts would get a default value of current timestamp or the time when the post call was made. In my analysis swagger editor shows the correct value of record_ts i.e current timestamp but its absent from the final json record. 

Any idea if this is expected?

Some potential solutions I am exploring:
VTL : http://velocity.apache.org/tools/2.0/apidocs/org/apache/velocity/tools/generic/DateTool.html
set default record timestamp in the aws request mapping stage and insert the data in kinesis firehose using datetime functions in VTL
Ref: https://forums.aws.amazon.com/thread.jspa?threadID=238417

AWS introduced Server Side Timestamp https://aws.amazon.com/about-aws/whats-new/2015/09/amazon-kinesis-server-side-timestamp/ 
But looks like this sol is only valid for Kinesis Stream and not Firehose (my use case). 


Thanks
Sid

**Comment by sidashah** [comment]
@sid88in
Use "additionalProperties" : false for that 
Ref: https://spacetelescope.github.io/understanding-json-schema/reference/object.html

**Comment by sid88in** [comment]
@sas2387 - Thanks for your input. I played around with additionalProperties in swagger but it doesn't seem to work. Here is the code I was testing (baseline: @rpgreen's example - https://gist.github.com/sid88in/a40927806a51c5c0509e57292690fc0f#file-additionalproperties-L88)

I am still able to post extra json fields. Also I came across these PR's 
https://github.com/OAI/OpenAPI-Specification/issues/668
https://github.com/mission-liao/pyswagger/issues/72

and https://support.reprezen.com/support/solutions/articles/6000162892-support-for-additionalproperties-in-swagger-2-0-schemas which talks about limited support for boolean values.

Does it work for you?

**Comment by sid88in** [comment]
@sas2387 - OK this works 
https://gist.github.com/sid88in/df498cd5b06b8a43610827e6809dc23e#file-this-works-L63

I basically focussed on additionalProperties and instead of passing a boolean variable, I am passing an empty schema reference (so ONLY json fields specified in swagger is allowed)

Now, above Test Case 1 and Test Case 2 gives me 400 (YAY! Thanks to you)
{
  "message": "Invalid request body"
}

Does this solution look OK? 

thanks a bunch
Sid

**Comment by sid88in** [comment]
@sas2387 - Looks like the above validation would fail in this case:
```[  
   { 
      "account-id":"abcdef123456",
      "type":"STOCK",
      "symbol":"AMZN",
      "shares":100,
      "details":{  
         "limit":1000,
         "temp": {
         	"something" : "something"
         }
      }
   }
]

OR

[  
   { 
      "account-id":"abcdef123456",
      "type":"STOCK",
      "symbol":"AMZN",
      "shares":100,
      "temp": {
         	"something" : "something"
      },      
      "details":{  
         "limit":1000
      }
   }
]

**Comment by Sunny19867** [comment]
Hi,
https://aws.amazon.com/blogs/compute/how-to-remove-boilerplate-validation-logic-in-your-rest-apis-with-amazon-api-gateway-request-validation/

Based on above, it seems schema validation can be only done for post requests. Is there any way to test "value" of query string parameter is of given format in GET? For example, if we are passing Date as query string parameter. We can say Date is required or not in "URL Query String Parameters". Can we say what format does the value need to be in? So I want to make sure Date is in format yyyy-MM-dd. 

Basically, can we do schema validation for GET Request? We got 
"Validate body, query string parameters, and headers" option available for Get method. Does that work?
We can use template method to create valid json using query string parameters. I was expecting request body will be validated with model schema based on body generated by template method using Get query parameters. If not, what is the intention of validating body for Get method?
